File: LeetCode Solutions/July 2022 LeetCoding Challenge/Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        head, current = None, root
        while(head != root):
            if(current.right == head):
                current.right = None
            if(current.left == head):
                current.left = None
            if(current.right):
                current = current.right
            elif(current.left):
                current = current.left
            else:
                current.right, head, current = head, current, root

File: LeetCode Solutions/May 2021 LeetCoding Challenge/Flatten Binary Tree to Linked List/Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        TreeNode head = null, current = root;
        while(head != root){
            if(current.right == head){
                current.right = null;
            }
            if(current.left == head){
                current.left = null;
            }
            if(current.right != null){
                current = current.right;
            }
            else if(current.left != null){
                current = current.left;
            } else{
                current.right = head;
                head = current;
                current = root;
            }
        }
    }
}